Rewrite conflicts annotation

peterjaap opened the pull request:
Magerun pull-request check-list:
- [X] Pull request against develop branch
- [X] readme.rst reflects changes 

Subject:    Check if one of the rewritten classes actually is written to fix the conflict.

Changes proposed in this pull request:

Check if one of the rewritten classes actually is written to fix the conflict in the list. 

When a fix for a merge conflict is introduced by creating another rewrite conflict, you can mark it as solved using an annotation. The annotation should be in the form (for example) `@SolvesRewriteConflict(catalog/product,First_ModuleWithConflict_Model_Catalog_Product, Second_ModuleWithConflict_Model_Catalog_Product, X_ModuleWithConflict_Model_Catalog_Product)`.

ktomk commented:
Hi Peter, can you share some context? Where does this annotation come from, how do users benefit from it or similar. Also if you have some link would be good to share here.

Build fails only for coverage ration if I see it right, just let me know if you've got issues to increase coverage then we can take a look together.

ktomk commented on src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php:
```diff
@@ -167,6 +171,48 @@ protected function _isInheritanceConflict($classes)
+                     */
+                    if (
+                        count($classesInAnnotation) > 1
+                        &&
```
You can merge this line with the next one, so it's one line with the operator in front. Same then for the next-next and next-next-next line :)

ktomk commented:
Please consider to squash your changes into one commit. You can force-push the feature branch then to update the pull-request.

peterjaap commented:
Thanks @ktomk, I never knew you could squash commits, awesome. Just squashed them into one.

The context is also in the PR message; magerun gives an overview of conflicting rewrites, like so;

```
$ n98-magerun dev:module:rewrite:conflicts
┌────────┬──────────────────────────────┬────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────┐
│models  │importexport/import_entity_pro│AvS_FastSimpleImport_Model_Import_Entity_Product,           │Inchoo_PHP7_Model_Import_Entity_Product                     │
│        │duct                          │Inchoo_PHP7_Model_Import_Entity_Product                     │                                                            │
└────────┴──────────────────────────────┴────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────┘
1 conflict was found!
```

Then, imagine somebody creates an extension that combines both modifications of those two rewrites into a new rewrite and making sure that rewrite is loaded first by letting those 2 extensions listed above be a dependency for the new one. It will then look like this;

```
$ n98-magerun dev:module:rewrite:conflicts
┌────────┬──────────────────────────────┬────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────┐
│models  │importexport/import_entity_pro│AvS_FastSimpleImport_Model_Import_Entity_Product,           │Inchoo_PHP7_Model_Import_Entity_Product, Elgentos_FixRewriteConflictBetweenAvSAndInchooPHP7_Model_Import_Entity_Product                     │
│        │duct                          │Elgentos_FixRewriteConflictBetweenAvSAndInchooPHP7_Model_Import_Entity_Product                     │                                                            │
└────────┴──────────────────────────────┴────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────┘
1 conflict was found!
```

But, although this technically still a rewrite conflict, we solved the issue of the collision because we merged the functionality in the eventual 'winner' of the rewrite. So we want to mark it as solved.

That's why I added the custom annotation `@SolvedRewriteConflict`. You add that annotation to the class that introduces the fix, so in this case `Elgentos_FixRewriteConflictBetweenAvSAndInchooPHP7_Model_Import_Entity_Product`.

So the annotation for the above example would be;
`@SolvesRewriteConflict(importexport/import_entity_product,AvS_FastSimpleImport_Model_Import_Entity_Product, Inchoo_PHP7_Model_Import_Entity_Product)`

Now we can add this to the code and have our deployment script run `magerun dev:module:rewrite:conflicts` to make the pre-deployment check pass instead of fail.

I hope this clarifies things.

tkn98 commented:
@peterjaap: Can you take a look to increase the test-coverage for your PR? Otherwise I left a little comment but overall code looks good in my eyes. A test should get things covered. If you have any issues with that, just let me know I can take care of it as well.

tkn98 commented on src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php:
```diff
@@ -167,6 +171,45 @@ protected function _isInheritanceConflict($classes)
+     * @var string $rewrittenClass
+     * @return bool
+     */
+    protected function _isARewriteFixingTheConflict($rewriteClasses, $rewrittenClass)
```
For new code we try to make the following differences:

# Prefer private over protected - do as little as necessary, this includes reduce visibility to the bare minimum.

# No underscore-prefix for method names - This is an outdated, very-old-PHP-4-style to signal method visibility. This is not needed any longer since PHP 5 nor recommended.

peterjaap commented:
Ok we can close this. I just confirmed this is actually unnecessary, as long as you fix the rewrite and **clear the cache**. Time wasted... well at least I learned how annotations work, haha :)

peterjaap commented:
I just followed the style as was already there in the class. Do you like me
to refactor that as well?
On Mon, 24 Oct 2016 at 13:26, Tom Klingenberg <notifications@github.com>
wrote:

> *@tkn98* commented on this pull request.
> ------------------------------
>
> In src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php
> <https://github.com/netz98/n98-magerun/pull/865#pullrequestreview-5429130>
> :
>
> > @@ -167,6 +171,45 @@ protected function _isInheritanceConflict($classes)
>      }
>
>      /**
> +     * Check if one of the rewritten classes actually is written to fix the conflict.
> +     * We can check by checking for an annotation in the PHPDOC block (@SolvesRewriteConflict)
> +     *
> +     * @var array $rewriteClasses
> +     * @var string $rewrittenClass
> +     * @return bool
> +     */
> +    protected function _isARewriteFixingTheConflict($rewriteClasses, $rewrittenClass)
>
> For new code we try to make the following differences:
> Prefer private over protected - do as little as necessary, this includes
> reduce visibility to the bare minimum. No underscore-prefix for method
> names - This is an outdated, very-old-PHP-4-style to signal method
> visibility. This is not needed any longer since PHP 5 nor recommended.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/netz98/n98-magerun/pull/865#pullrequestreview-5429130>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAaVAHDkMcfNrQIqyHTPH8PsLwy3-bZEks5q3JXhgaJpZM4Ka_rq>
> .
>

tkn98 commented:
@peterjaap: Not as you wrote it's closed now (for a reason). Thanks again for your efforts.

peterjaap commented:
I just ran into a situation where this is useful; I fixed a rewrite conflict with a new extension, but did not change the XML of the original extension since its pulled in via Composer. This annotation will allow you to mark it as solved. Magerun will now show this;

```
├────────┼──────────────────────────────┼────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────┤
│models  │customer/customer             │Shopliebe_PasswordStrength_Model_Customer,                  │Elgentos_ExtendedPasswordStrength_Model_Customer_Customer   │
│        │                              │EW_NativePasswords_Model_Customer_Customer,                 │                                                            │
│        │                              │Elgentos_ExtendedPasswordStrength_Model_Customer_Customer   │                                                            │
└────────┴──────────────────────────────┴────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────┘
```

But when annotated with;

```
/**
    @SolvesRewriteConflict(customer/customer,Shopliebe_PasswordStrength_Model_Customer, EW_NativePasswords_Model_Customer_Customer)
 */
```

It shouldn't show that message since it is marked as fixed. I will fix my commit once I have time.
